### PR TITLE
perf(mlx): fuse recurrent decode convolution and SiLU

### PR DESCRIPTION
**Pull request by Lyxot**
Recurrent single-token decode currently materializes a depthwise convolution result before applying SiLU. Fuse those operations in a Metal kernel while preserving the native accumulation, cast, and activation order. The scoped adapter recognizes the native source contract rather than a model-name allowlist, and restores instance classes on exit. Prefill, training, unsupported shapes or dtypes, distributed models, and changed upstream contracts keep native behavior.

Eight native Metal tests passed after rebasing. Coverage checks exact numerics for supported BF16/FP16 inputs, native fallbacks, source-contract refusal, and restoration. HTTP measurements use this branch alone; Qwen 0.8B has 18 eligible recurrent layers and Gemma E2B has none, providing a no-op control.

```bash
python -m pytest tests/test_mlx_decode_metal.py -q
```

### HTTP benchmark

Six fixed paired repetitions per case, balanced AB/BA; 512 generated tokens, greedy seed 42, identical snapshot/native context and unquantized KV within pairs. M3 Max / 128 GiB, MLX 0.32.2, mlx-lm 0.31.3, mlx-vlm 0.7.0, Transformers 5.5.0. Loading, unloading and generation all use HTTP. Warmups and conditioning are excluded; MoE variants load one at a time. Qwen 0.8B is BF16, Gemma E2B mixed 4/8-bit QAT, both large MoEs 8-bit.

| Case | Metric | Valid pairs | Median baseline → candidate | Paired change (95% CI) | CV baseline / candidate | Result |
|---|---|---:|---:|---:|---:|---|
| decode-gemmae2 | decode (tokens/s) | 5/6 | 101.95 → 102.07 | -0.04% (-0.30% to 0.22%) | 0.31% / 0.28% | Stable; no improvement resolved |
| decode-qwen08 | decode (tokens/s) | 6/6 | 157.52 → 162.58 | 3.20% (3.05% to 3.34%) | 0.16% / 0.07% | Stable improvement |

Changes are geometric mean paired speed ratios with Student-t 95% intervals on paired log ratios. Stable requires ≥4 valid pairs and both variant CV ≤1%; improvement additionally requires a positive lower confidence bound. No throughput-based rejection or adaptive extension. Intervals describe this session, not cross-day reproducibility.

Physical checks: nominal OS thermal pressure, ≤75°C admission and ≤3°C paired start difference, zero new global pageouts/swapouts, equal output text/token counts. Prefill and decode independently require phase-clock matching/drift ≤1% and telemetry coverage ≥90%. Prefill telemetry includes HTTP/CPU/first-token overhead. Warm TTFT is too short for reliable per-phase clock sampling and uses shared checks only. Matching GPU clocks does not prove an otherwise idle system.

Copy uses an 8K reused prefix; its primary metric is warm TTFT (positive speed ratio means faster, not the same percentage as latency reduction). Prefill/MoE/integration use 4K cold prompts (the explicitly labelled Gemma 8K prefill case is a separate longer-window experiment); recurrent decode 1K. Zoo comparisons hold Studio integration constant, with only the named Zoo optimization present. Integration alone uses baseline Zoo; Gemma E2B decode is a no-op control. Gains are not additive.

The three Zoo optimization PRs are independently based on main and introduce the same `inference.py` module; merging multiple requires combining their exports. Studio activation is provided by https://github.com/unslothai/unsloth/pull/10733.

<details>
<summary>All measured primary-metric pairs and exclusions</summary>

| Case / metric | A→B values (tokens/s; TTFT in seconds) | Accepted | Reasons |
|---|---|---|---|
| decode-gemmae2 / decode / 1 (AB) | 102.2611 → 102.0722 | yes | — |
| decode-gemmae2 / decode / 2 (BA) | 101.9529 → 102.2739 | yes | — |
| decode-gemmae2 / decode / 3 (AB) | 101.8299 → 101.7510 | yes | — |
| decode-gemmae2 / decode / 4 (BA) | 102.6278 → 102.4272 | yes | — |
| decode-gemmae2 / decode / 5 (AB) | 101.7771 → 101.7018 | yes | — |
| decode-gemmae2 / decode / 6 (BA) | 101.1912 → 102.1648 | no | temperature-gap |

decode-gemmae2 decode: all-pair estimate 0.12% (95% CI -0.35% to 0.60%; baseline/candidate CV 0.43% / 0.26%); accepted AB -0.11%, BA 0.06%.

| Case / metric | A→B values (tokens/s; TTFT in seconds) | Accepted | Reasons |
|---|---|---|---|
| decode-qwen08 / decode / 1 (AB) | 157.7860 → 162.7150 | yes | — |
| decode-qwen08 / decode / 2 (BA) | 157.2319 → 162.3702 | yes | — |
| decode-qwen08 / decode / 3 (AB) | 157.3256 → 162.5891 | yes | — |
| decode-qwen08 / decode / 4 (BA) | 157.7104 → 162.4915 | yes | — |
| decode-qwen08 / decode / 5 (AB) | 157.3180 → 162.5720 | yes | — |
| decode-qwen08 / decode / 6 (BA) | 157.8529 → 162.7056 | yes | — |

decode-qwen08 decode: all-pair estimate 3.20% (95% CI 3.05% to 3.34%; baseline/candidate CV 0.16% / 0.07%); accepted AB 3.27%, BA 3.12%.

</details>

<details>
<summary>Checkpoint revisions and context limits</summary>

| Alias | Model | Revision | Checkpoint dtype | Context |
|---|---|---|---|---:|
| gemmae2 | mlx-community/gemma-4-E2B-it-qat-4bit | `42f62737af7a9fd8c1d55d79666c1a217be4e2e2` | QAT mixed 4/8-bit | 131072 |
| qwen08 | Qwen/Qwen3.5-0.8B | `2fc06364715b967f1860aea9cf38778875588b17` | BF16/F32 native | 262144 |

</details>

Measured candidate `cf5788ab6924008c849ab577e17883b31a6a3c8b` against `a7eadfb1c16532b78b5fd8ea1a8bf91d635d71cf`.

### CODE REVIEW

**Comment by Lyxot** [comment]
@codex review

Please review the current head. Scope: Recurrent decode convolution and SiLU fusion only.

Report any major issues you find. If you find none, please post a normal PR comment saying "Didn't find any major issues", rather than only a reaction or activity-summary update.

**Comment by danielhanchen** [comment]
The measurements look careful, but unsloth_zoo/mlx/inference.py reproduces and re-execs upstream __call__ and _causal_conv1d_decode source through AST matching, which will silently stop applying whenever mlx-vlm edits those functions. Could you narrow it to a plain subclass or method override that calls the Metal kernel, or push the fusion upstream, for a 3% decode gain?

**Comment by danielhanchen** [comment]
The paired numbers are convincing on the Qwen recurrent path, but the head still rewrites upstream __call__ and _causal_conv1d_decode from their parsed source, so it will quietly fall back the moment mlx-vlm edits those bodies. Could you narrow it to a subclass or method override that calls the Metal kernel directly, or take the fusion upstream?

**Comment by danielhanchen** [comment]
The head still rebuilds upstream `__call__` and `_causal_conv1d_decode` through AST rewriting and exec in unsloth_zoo/mlx/inference.py, which is the part I asked about earlier. Could you replace that with a plain subclass method that calls the Metal kernel, or take the fusion upstream to mlx-vlm?

**Comment by danielhanchen** [comment]
Both red checks here are one thing, and it is this PR's rather than the flakiness going around yesterday. Reproduced against current `main` merged with `688d848a`:

```
2 dynamic-execution call site(s) not in the baseline:

  unsloth_zoo/mlx/inference.py:154  compile(...) is handed a value that is not written out
  unsloth_zoo/mlx/inference.py:154  exec(...) is handed a value that is not written out
```

That is `Source lint`, and `Repo tests (CPU)` fails on the same fact through `tests/security/test_lint_exec_literals.py::test_the_baseline_matches_the_tree_it_was_recorded_against`. Everything else passes: ruff, compileall, the dynamic-exec lint, the trigger lint, and `tests/test_mlx_decode_metal.py` skips cleanly on a Linux runner with no MLX.

The gate is doing its job. `_function_from_ast` execs a code object built at runtime, so it has to be recorded as reviewed:

```
python scripts/lint_exec_literals.py --update
```

then replace the `"REVIEW ME"` reason the update writes in `scripts/exec_literals_baseline.json`. `--update` alone leaves the gate red on purpose, so the reason is not optional.

For what it is worth the justification looks straightforward to me: the tree you compile comes from `ast` over `base.__call__` and `base._causal_conv1d_decode`, which is this repo's own installed source, and `_CONV_SILU_CONTRACT` pins the upstream it is allowed to rewrite. Nothing downloaded or caller-supplied reaches it. Worth saying that explicitly in the reason field, since that is the distinction the baseline exists to record.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex security review

**Comment by danielhanchen** [comment]
Pushed two commits to this branch, both of which were blocking rather than optional.

`557c2d6f` records the AST recompile in `scripts/exec_literals_baseline.json`. `Source lint` and the `tests/security` hard gate were both red on this branch (`unsloth_zoo/mlx/inference.py:154  exec(...) is handed a value that is not written out`, verdict `new-call-site`), and `--update` alone leaves them red because it writes a `REVIEW ME` reason that the gate then rejects, so the justification has to be written by hand. Both gates are green with it.

`cfc018c3` gives the new module the zoo's AGPL header. Every other file under `unsloth_zoo/` carries the full block; the SPDX one-liner is the `studio/` subtree's convention.

Then I ran it on real Apple Silicon, which had never happened on this PR: `MLX CI on Mac M1` is behind the `mlx` label and reported `skipping` on every push. I added the label, and replicated the branch onto a staging repo to run both Apple images.

**macos-15 and macos-26**, mlx 0.32.1 / mlx-lm 0.31.3 / mlx-vlm 0.6.4 / transformers 5.5.0, Metal confirmed live (a custom `mx.fast.metal_kernel` probe runs before anything else):

- `tests/test_mlx_decode_metal.py`: 12 passed, 0 skipped, on both images.
- `mlx-community/Qwen3.5-0.8B-4bit`, its own 18 recurrent layers driven through prefill(3) then two decode(1) steps with a real `ArraysCache`, inside and outside the scope: 18 of 18 layers patched, 36 fused kernel calls during decode (two per layer), 0 during prefill, 0 bitwise differences in outputs or cache state, all classes restored on exit.
- `mlx-community/SmolLM-135M-Instruct-4bit` as a non-recurrent control: 0 layers patched, 0 kernel calls.

So the kernel does what it says on real hardware, and `compile_options={"math_mode": "safe"}` is accepted by the pinned mlx 0.32.1, not only by 0.32.2.

Two observations, neither blocking.

The eligible class depends on which backend built the model, and there is no diagnostic when it does not match. `mlx-community/Qwen3.5-0.8B-4bit` carries a `vision_config`, so `_is_vlm` is true and unsloth builds it through mlx-vlm, whose `Qwen3_5GatedDeltaNet` exposes `_causal_conv1d_decode`. Loading the same files with `mlx_lm.load` yields mlx-lm's own `GatedDeltaNet`, which has no such method, so `_decode_conv_silu_contract` returns `None` at its first guard and nothing is patched. That cost me a full debugging cycle, because the result is zero eligible layers, zero kernel calls and identical output, with nothing logged. A one-shot debug line naming the class that was examined and rejected would make it self-diagnosing.

`fused_decode_conv_silu` has no reference count, unlike `fused_moe_gate_up`'s `active_scopes` in #1189. Enter scope A, enter scope B over the same module, exit A: the class is restored while B is still open, and B silently runs native for the rest of its life. Numerics stay correct because the fallback is the native path, so this is a performance only defect, and with unslothai/unsloth#10733 entering the decode scope per generation it is nested rather than overlapping today. It becomes reachable the moment two generations share one model, which is the same case #1189 already answers.

Also worth noting that the module level `pytestmark = skipif(not mx.metal.is_available())` means this file runs 0 of 12 tests off Apple Silicon, including the seven that only assert contract refusal and never dispatch a kernel. I ran those seven by hand on the Linux CPU backend and all seven pass, so they would gate for free on the ubuntu runner.

File: unsloth_zoo/mlx/inference.py [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+    def conv_silu(self, x, weight):
+        if (x.ndim == 3 and x.shape[0] > 0 and 2 <= x.shape[1] <= 8 and x.shape[2] > 1
+                and x.dtype in (mx.bfloat16, mx.float16)
+                and weight.shape == x.shape[1:] and weight.dtype == mx.float32):
+            return _decode_conv_silu(x, weight)
+        return silu(conv(x, weight))
+
+    def fused_call(self, *args, **kwargs):
+        if (self.training or prepare.__globals__.get(conv_name) is not conv
+                or not _bindings_intact(bindings)):
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/inference.py:448]: Checked this against the code rather than the description. _function_from_ast rebuilds the adapted method with FunctionType(..., original.__globals__, ...), so its globals ARE the language module's live dict: I rebound native.mx mid-scope and the adapted body picked up the new object, exactly as the native method would. The divergence is therefore confined to the one rewritten expression, whose conv and silu are captured, and it needs a caller to rebind a third-party module's mx or nn globals while a generation scope is open. _decode_conv_silu_contract already pins both aliases at patch time, and the call-time guard covers the conv binding plus the nn.silu fingerprint. Below the bar for a change here: not reachable on real inference input, and rechecking every global alias on every decode step costs more than the drift it would catch.

**Comment by danielhanchen** [comment]
@codex review

File: tests/test_mlx_decode_metal.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+import numpy as np
+import pytest
+
+mx = pytest.importorskip("mlx.core")
+from mlx_simulation import mlx_is_simulated
+
+if mlx_is_simulated():
+    pytest.skip("Requires native MLX", allow_module_level = True)
+pytestmark = pytest.mark.skipif(not mx.metal.is_available(), reason = "Requires Metal")
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tests/test_mlx_decode_metal.py]: Correct, and it matches what I hit running this by hand: the file executes nowhere today, which is why the only Metal evidence for it so far came from a staging workflow I wrote. Fixed in e0785be7 by adding the suite to the mlx-smoke step list in .github/workflows/mlx-ci.yml, beside the attention kernel suite. The lane is opt-in behind the mlx label, so I labelled this PR too; on the label-triggered run it was 12 passed, 0 skipped on both macos-15 and macos-26.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/mlx/inference.py [hunk 3]
```diff
@@ -284,3 +302,188 @@ def fused_moe_gate_up(model):
...
+
+    return type(f"_FusedDecodeConvSiLU{base.__name__}", (base,), {
+        "__call__": fused_call,
+        "_unsloth_decode_conv_silu": _function_from_ast(prepare, inner),
+        "_unsloth_apply_conv_silu": conv_silu,
+    })
+
+
+@contextmanager
+def fused_decode_conv_silu(model):
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/inference.py]: Fixed in c9cee6168. Entered beside fused_moe_gate_up in generation_mode and in both loader generation paths, so the kernel now runs for library users and not only for callers that build their own scope.

That made overlapping ownership routine, which the scope could not survive: the first exit restored the native class while a later scope still held the module, silently losing the kernel for the rest of that scope. Owners are counted now and only the last one restores. Regression test: test_overlapping_scopes_keep_the_module_fused_until_the_last_exit.

One more thing the wiring exposed: the membership tests read a module's own dict entries, and generation hands these scopes whatever named_modules() yields, including plain stand-ins, which raised TypeError out of the generation path. Non-Module objects are skipped now, covered by test_the_fusion_scopes_tolerate_whatever_named_modules_yields.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/mlx/inference.py [hunk 4]
```diff
@@ -284,3 +302,214 @@ def fused_moe_gate_up(model):
...
+        for module in reversed(changed):
+            scopes = getattr(module, "_unsloth_decode_scopes", 0)
+            if scopes > 1:
+                module._unsloth_decode_scopes = scopes - 1
+                continue
+            if type(module) is getattr(module, "_unsloth_decode_patched", None):
+                module.__class__ = module._unsloth_decode_native
+            for name in ("_unsloth_decode_scopes", "_unsloth_decode_native",
+                         "_unsloth_decode_patched"):
+                module.__dict__.pop(name, None)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/inference.py:515]: This one does not hold. mlx routes by value type, not by attribute:

    def __setattr__(self, key, val):
        if isinstance(val, (mx.array, dict, list, tuple)):
            ...
            self[key] = val
        else:
            super(Module, self).__setattr__(key, val)
            self.pop(key, None)

Only arrays and containers become mapping entries. Everything else goes to __dict__, and mlx even pops the mapping key to keep the two from disagreeing. The three values here are an int and two classes, so none of them is a mapping entry, and __dict__.pop is what clears them. module.pop would remove nothing.

Measured on mlx 0.32.1 with the scope actually fused:

    fused: True
      __dict__ : ['_unsloth_decode_native', '_unsloth_decode_patched', '_unsloth_decode_scopes']
      mapping  : []

and after the exit both are empty and the mapping is byte-for-byte the set it started with. This is also why the membership tests above read the mapping: they are looking for an instance entry that shadows the class method, which is a different question from where a plain attribute lands.

Taking the last half of the suggestion though, since it costs nothing and guards a real future hazard: 9a5b3d3e0 asserts the mapping is clean as well, so a value that later starts landing there cannot leak past the exit unnoticed.

**Comment by danielhanchen** [comment]
@codex review
